assets/src/asset.rs

```diff
@@ -0,0 +1,166 @@
+    /// The extension of this asset
+    pub ext: &'static str,
+    /// The name of this asset.
+    pub name: String,
+    /// The storage id, uniquely identifying the storage it's been loaded from.
+    pub store: StoreId,
+}
+
+impl AssetSpec {
+    /// Creates a new asset specifier from the given parameters.
+    pub fn new(name: String, ext: &'static str, store: StoreId) -> Self {
+        AssetSpec {
+            ext,
+            name,
+            store,
+        }
+    }
+}
+
+/// A basic implementation for a cache. This might be useful as the `Context` of
+/// an `Asset`, so that the same asset doesn't get imported twice.
+///
+/// Because contexts have to be immutable, a `RwLock` is used. Therefore, all
+/// operations are blocking (but shouldn't block for a long time).
+pub struct Cache<T> {
+    map: RwLock<FnvHashMap<AssetSpec, T>>,
+}
+
+impl<T> Cache<T>
+    where T: Clone
+{
+    /// Creates a new `Cache` and initializes it with the default values.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Inserts an asset, locking the internal `RwLock` to get write access to the hash map.
+    ///
+    /// Returns the previous value in case there was any.
+    pub fn insert(&self, spec: AssetSpec, asset: T) -> Option<T> {
+        self.map.write().insert(spec, asset)
+    }
+
+    /// Retrieves an asset, locking the internal `RwLock` to get read access to the hash map.
+    /// In case this asset has been inserted previously, it will be cloned and returned.
+    /// Otherwise, you'll receive `None`.
+    pub fn get(&self, spec: &AssetSpec) -> Option<T> {
+        self.map.read().get(spec).map(Clone::clone)
+    }
+
+    /// Deletes all cached values, except the ones `f` returned `true` for.
+    /// May be used when you're about to clear unused assets (see `Asset::clear`).
+    ///
+    /// Blocks the calling thread for getting write access to the hash map.
+    pub fn retain<F>(&self, f: F)
+        where F: FnMut(&AssetSpec, &mut T) -> bool
+    {
+        self.map.write().retain(f);
+    }
+
+    /// Deletes all cached values after locking the `RwLock`.
+    pub fn clear_all(&self) {
+        self.map.write().clear();
+    }
+}
+
+impl<T> Default for Cache<T> {
+    fn default() -> Self {
+        Cache { map: Default::default() }
+    }
+}
+
+/// A format, providing a conversion from bytes to asset data, which is then
+/// in turn accepted by `Asset::from_data`. Examples for formats are
+/// `Png`, `Obj` and `Wave`.
+pub trait Format
+    where Self: Sized
+{
+    /// The data type this format is able to load.
+    type Data;
+    /// The kind of error it may produce.
+    type Error: Error;
+
+    /// Returns the extension (without `.`).
+    ///
+    /// ## Examples
+    ///
+    /// * `"png"`
+    /// * `"obj"`
+    /// * `"wav"`
+    fn extension() -> &'static str;
+
+    /// Reads the given bytes and produces asset data.
+    fn parse(&self, bytes: Vec<u8>) -> Result<Self::Data, Self::Error>;
```

Comment:
There was a discussion on this but Github marked it outdated due to other changes.

No it would not be better to take a slice, primarily because the bytes would be dropped after this function is called and for some very simple asset types the bytes themselves might be part of the asset, so passing in a Vec here allows us to save an allocation in some instances, and is harmless in others.